### PR TITLE
Feat: OAuth authorization server metadata

### LINKED ISSUE #46: Support for RFC 8414 - OAuth Authorization Server Metadata
See: https://github.com/mastodon/mastodon/pull/29191

- [x] `GET /.well-known/oauth-authorization-server`
- [x] Correct URL generation for endpoints
- [x] Return values in scopes Enum `scopeEnum.enumValues`
- [x] `response_types_supported` of `code`
- [x] `response_modes_supported` of `query`
- [x] `grant_types_supported` of `authorization_code` (I'm not sure if we really support `client_credentials`)
- [x] `token_endpoint_auth_methods_supported` of `client_secret_post` (we don't support `client_secret_basic`)

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
- Fixes #46 
- Also fixes an issue around `redirect_uri` being required for `client_credentials` grant flow (it's actually not required for that flow)
- Ensures that the `redirect_uri` in the token request at least is present within the registered application — however, if we moved to Access Grants being stored, then we'd check "does this access grant authorize redirecting to this URI?" which is actually the OAuth logic here: if you start an OAuth flow with authorize using redirectUri1 but then try to exchange the code with redirectUri2, it should fail.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new endpoint for OAuth server discovery at `/.well-known/oauth-authorization-server`.
	- Enhanced OAuth functionality with improved error handling and validation for token requests.
	- Added support for two-factor authentication.
	- Introduced functionality for reporting remote accounts and posts.
	- Improved alignment with Mastodon API changes regarding OAuth and applications.
	- Added support for RFC 8414 for OAuth Authorization Server metadata endpoint.

- **Bug Fixes**
	- Improved error messages for invalid redirect URIs and missing required fields in token requests.

### CODE REVIEW

File: src/oauth.tsx [hunk 1]
```diff
@@ -322,7 +322,7 @@ const tokenRequestSchema = z.object({
   code: z.string().optional(),
   client_id: z.string(),
   client_secret: z.string(),
-  redirect_uri: z.string().url(),
+  redirect_uri: z.string().url().optional(),
```
Comment (ThisIsMissEm) [inline src/oauth.tsx:325]: Another option might be a discriminated union, so "these are the parameters required for authorization_code and these are required for client_credentials"

Technically `client_secret` is also optional once you support public clients, not just confidential clients.

[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply src/oauth.tsx:325]: @dahlia thoughts on using `superRefine`? one issue is that the errors returned from Zod aren't actually valid OAuth 2 error responses.

[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply src/oauth.tsx:325]: Like:
```
{
  "error": "Invalid request",
  "zod_error": {
    "issues": [
      {
        "expected": "'authorization_code' | 'client_credentials'",
        "received": "undefined",
        "code": "invalid_type",
        "path": [
          "grant_type"
        ],
        "message": "Required"
      }
    ],
    "name": "ZodError"
  }
}
```

Doesn't conform to:

```
interface ErrorResponse {
  error: string;
  error_description?: string;
  error_uri?: string;
}
```

Per https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

**Comment by ThisIsMissEm** [comment]
> @ThisIsMissEm Oh, could you add an entry to the _CHANGES.md_ file?

Done.

File: src/oauth.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply src/oauth.tsx:325]: Conforming to OAuth 2 error responses sounds more important to me.

File: src/oauth.tsx [hunk 2]
```diff
@@ -444,4 +470,26 @@ app.post("/token", cors(), async (c) => {
...
+    response_modes_supported: ["query"],
+    grant_types_supported: ["authorization_code", "client_credentials"],
+    token_endpoint_auth_methods_supported: [
+      "client_secret_post",
+      // Not supported by Hollo:
+      // "client_secret_basic",
+    ],
+    app_registration_endpoint: new URL("/api/v1/apps", url).href,
+  });
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/oauth.tsx:493]: I think it's okay and actually it's more consistent with the existing code.  As we have `BEHIND_PROXY` configuration we don't have to worry about it.

**Pull request merged**